File: BirdProcessor/WordProcessor/WordProcessor.m
#import "WordProcessor.h"

@interface WordProcessor ()

@property (nonatomic) id<StringManipulatorProtocol> stringManipulator;

@end

@implementation WordProcessor

- (instancetype)initWithStringManipulator:(id<StringManipulatorProtocol>)stringManipulator {
    self = [super init];
    if (self) {
        self.stringManipulator = stringManipulator;
    }
    return self;
}

- (NSString *)processInputString:(NSString *)string {
    if (self.stringManipulator == nil) {
        return @"stringManipulator is nil!";
    }
    
    BOOL isEmpty = [self.stringManipulator isEmptyString:string];
    NSString *isEmptyResult = isEmpty ? @"is" : @"is not";
    
    NSInteger numberOfWords = [self.stringManipulator numberOfWordsInString:string];
    NSString *numberOfWordsResult = [NSString stringWithFormat:@"%ld word(s) and ", numberOfWords];
    
    NSInteger numberOfCharacters = [self.stringManipulator numberOfCharactersInString:string];
    NSString *numberOfCharactersResult = [NSString stringWithFormat:@"%ld character(s)", numberOfCharacters];
    
    return [NSString stringWithFormat:@"The string %@ empty. It has %@%@.",
            isEmptyResult,
            numberOfWordsResult,
            numberOfCharactersResult];
}

@end

File: BirdProcessor/WordProcessor/WordProcessor.swift
// Because this class is used in Objective-C, we still need `@objc` and `NSObject`:
@objc class WordProcessor: NSObject {
    
    private let stringManipulator: any StringManipulatorProtocol
    
    @objc weak var delegate: (any WordProcessorDelegate)?
    
    @objc init(stringManipulator: any StringManipulatorProtocol) {
        self.stringManipulator = stringManipulator
    }
    
    @objc func processInputString(_ string: String) -> String {
        delegate?.wordProcessor(self, didBeginProcessingInput: string)
        // This runs *after* the resulting string is constructed
        // but *before* control is returned to the caller—
        // in other words, when processing is truly finished
        defer { delegate?.wordProcessor(self, didFinishProcessingInput: string) }
        
        let isEmpty = stringManipulator.isEmpty(string)
        let isEmptyResult = isEmpty ? "is" : "is not"
        
        // Note the ? before the arguments:
        let numberOfWordsResult = if let numberOfWords = stringManipulator.numberOfWords?(in: string) {
            "\(numberOfWords) word(s) and "
        } else { "" }
        
        if numberOfWordsResult.isEmpty {
            delegate?.wordProcessor(self, didEncounterError: "Cannot count words")
        }
        
        let numberOfCharacters = stringManipulator.numberOfCharacters(in: string)
        let numberOfCharactersResult = "\(numberOfCharacters) character(s)"
        
        return "The string \(isEmptyResult) empty. It has \(numberOfWordsResult)\(numberOfCharactersResult)."
    }
    
}
